packages/vega-parser/src/parsers/guides/axis-grid.js

```diff
@@ -38,49 +40,80 @@ export default function(spec, config, userEncode, dataRef, band) {
     round:  _('tickRound')
   };
 
-  if (orient === Top || orient === Bottom) {
-    u = 'x';
-    v = 'y';
-    s = 'height';
-  } else {
-    u = 'y';
-    v = 'x';
-    s = 'width';
-  }
-  v2 = v + '2';
-
-  update[u] = enter[u] = exit[u] = tickPos;
-
-  if (vscale) {
-    update[v] = enter[v] = {scale: vscale, range: 0, mult: sign, offset: offset};
-    update[v2] = enter[v2] = {scale: vscale, range: 1, mult: sign, offset: offset};
+  if (!isSignal(orient)) {
+    if (orient === Top || orient === Bottom) {
+      u = 'x';
+      v = 'y';
+      s = 'height';
+    } else {
+      u = 'y';
+      v = 'x';
+      s = 'width';
+    }
+    v2 = v + '2';
+  
+    update[u] = enter[u] = exit[u] = tickPos;
+  
+    if (vscale) {
+      update[v] = enter[v] = {scale: vscale, range: 0, mult: sign, offset: offset};
+      update[v2] = enter[v2] = {scale: vscale, range: 1, mult: sign, offset: offset};
+    } else {
+      update[v] = enter[v] = {value: 0, offset: offset};
+      update[v2] = enter[v2] = {signal: s, mult: sign, offset: offset};
+    }
   } else {
-    update[v] = enter[v] = {value: 0, offset: offset};
-    update[v2] = enter[v2] = {signal: s, mult: sign, offset: offset};
+    update.x = enter.x = xAxisConditionalEncoding(orient.signal, tickPos, vscale ? { scale: vscale, range: 0, mult: sign, offset: offset} : { value: 0, offset: offset});
+    update.y = enter.y = xAxisConditionalEncoding(orient.signal, tickPos, vscale ? { scale: vscale, range: 0, mult: sign, offset: offset} : { value: 0, offset: offset}, false);
+    exit.x = xAxisConditionalEncoding(orient.signal, tickPos, null);
+    exit.y = xAxisConditionalEncoding(orient.signal, tickPos, null, false);
+    update.y2 = enter.y2 = xAxisConditionalEncoding(orient.signal, vscale ? {scale: vscale, range: 1, mult: sign, offset: offset} : {signal: xAxisExpr(orient.signal, 'height', 'width', false, false).signal, mult: sign, offset: offset}, null);
+    update.x2 = enter.x2 = xAxisConditionalEncoding(orient.signal, vscale ? {scale: vscale, range: 1, mult: sign, offset: offset} : {signal: xAxisExpr(orient.signal, 'height', 'width', false, false).signal, mult: sign, offset: offset}, null, false);
   }
-
+  
   return guideMark(RuleMark, AxisGridRole, null, Value, dataRef, encode, userEncode);
 }
 
 function offsetValue(offset, sign)  {
-  if (sign === 1) {
-    // do nothing!
-  } else if (!isObject(offset)) {
-    offset = sign * (offset || 0);
+  var entry;
+  if (!isSignal(sign)) {
```

Comment:
Instead of breaking by `isSignal(sign)` at the top-level branch, I think you can keep a lot of existing structure and check if `isSignal(sign)` in the existing if/else and while loop structure. 

This will avoid redundant code and make it easier to assess if we're dealing with `sign` signal correctly.

Reply:
refactored

Reply:
Why is this `!isSignal(sign)` still here?

Reply:
removed now